# Exploring India's Consumer Price Index (CPI) Data with Pandas

**Dataset:** State/UT-wise General Consumer Price Index and Inflation, July 2026 (Provisional)  
**Source:** Government of India CPI press-release annexure  
**Goal:** Learn how to inspect, clean, filter, sort, summarise and save a real-world government dataset using Pandas.

In [36]:
pip install pandas openpyxl


[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


## 1. Import Pandas

In [2]:
import pandas as pd
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## 2. Load the Excel file

`pd.read_excel()` reads an Excel file and stores it as a Pandas dataframe.

In [3]:
df = pd.read_excel("cpi_july26.xlsx", header=None)

In [4]:
df.head(10)

,0,1,2,3,4,5,6,7
0,Annexure-III,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,State/UT wise general index and inflation in J...,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,S.No.,State,CPI,NaN,NaN,Inflation,NaN,NaN
3,NaN,NaN,Rural,Urban,Combined,Rural,Urban,Combined
4,1.,Andaman And Nicobar Islands,111.13,107.44,109.32,4.99,3.26,4.15
5,2.,Andhra Pradesh,107.81,107.55,107.71,6.13,5.09,5.72
6,3.,Arunachal Pradesh,105.13,105.09,105.12,2.48,2.8,2.56
7,4.,Assam,106.03,104.45,105.71,3.74,2.53,3.49
8,5.,Bihar,107.01,105.86,106.85,3.72,2.84,3.59
9,6.,Chandigarh*,0,108.04,108.04,0,4.3,4.3


## 3. Inspect the raw dataframe

In [5]:
df.shape

(41, 8)

In [6]:
df.tail(5)

,0,1,2,3,4,5,6,7
36,33.,Tripura,104.55,104.75,104.59,2.11,3.04,2.32
37,34.,Uttar Pradesh,108.92,107.58,108.5,5.25,3.62,4.74
38,35.,Uttarakhand,109.41,106.75,108.44,4.66,3.43,4.21
39,36.,West Bengal,107.39,106.76,107.11,4.2,3.5,3.89
40,*: No rural market in Chandigarh,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 41 entries, 0 to 40
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       40 non-null     str   
 1   1       37 non-null     str   
 2   2       38 non-null     object
 3   3       37 non-null     object
 4   4       37 non-null     object
 5   5       38 non-null     object
 6   6       37 non-null     object
 7   7       37 non-null     object
dtypes: object(6), str(2)
memory usage: 2.7+ KB


In [8]:
cpi = df.iloc[4:40].copy()
cpi.columns = [
    "s_no", "state", "cpi_rural", "cpi_urban", "cpi_combined",
    "inflation_rural", "inflation_urban", "inflation_combined"
]
cpi.head()

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
4,1.,Andaman And Nicobar Islands,111.13,107.44,109.32,4.99,3.26,4.15
5,2.,Andhra Pradesh,107.81,107.55,107.71,6.13,5.09,5.72
6,3.,Arunachal Pradesh,105.13,105.09,105.12,2.48,2.8,2.56
7,4.,Assam,106.03,104.45,105.71,3.74,2.53,3.49
8,5.,Bihar,107.01,105.86,106.85,3.72,2.84,3.59


## Clean the state names

In [9]:
cpi["state"] = cpi["state"].astype(str).str.strip()
cpi["state"].tolist()

['Andaman And Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chandigarh*',
 'Chhattisgarh',
 'Goa',
 'Gujarat',
 'Haryana',
 'Himachal Pradesh',
 'Jammu And Kashmir',
 'Jharkhand',
 'Karnataka',
 'Kerala',
 'Ladakh',
 'Lakshadweep',
 'Madhya Pradesh',
 'Maharashtra',
 'Manipur',
 'Meghalaya',
 'Mizoram',
 'Nagaland',
 'NCT of Delhi',
 'Odisha',
 'Puducherry',
 'Punjab',
 'Rajasthan',
 'Sikkim',
 'Tamil Nadu',
 'Telangana',
 'The Dadra And Nagar Haveli And Daman And Diu',
 'Tripura',
 'Uttar Pradesh',
 'Uttarakhand',
 'West Bengal']

## Convert numeric columns

In [10]:
numeric_columns = [
    "s_no", "cpi_rural", "cpi_urban", "cpi_combined",
    "inflation_rural", "inflation_urban", "inflation_combined"
]
for column in numeric_columns:
    cpi[column] = pd.to_numeric(cpi[column], errors="coerce")
cpi.dtypes

s_no                  float64
state                     str
cpi_rural             float64
cpi_urban             float64
cpi_combined          float64
inflation_rural       float64
inflation_urban       float64
inflation_combined    float64
dtype: object

## Check for missing values

In [11]:
cpi.isnull().sum()

s_no                  36
state                  0
cpi_rural              0
cpi_urban              0
cpi_combined           0
inflation_rural        0
inflation_urban        0
inflation_combined     0
dtype: int64

The source notes that Chandigarh has no rural market. Its rural CPI and rural inflation values are therefore shown as zero in the source and should not be interpreted as a normal measured rural observation.

## Check for duplicate rows

In [12]:
cpi.duplicated().sum()

np.int64(0)

## Check the cleaned dataset

In [13]:
cpi.shape

(36, 8)

In [14]:
cpi.head()

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
4,NaN,Andaman And Nicobar Islands,111.13,107.44,109.32,4.99,3.26,4.15
5,NaN,Andhra Pradesh,107.81,107.55,107.71,6.13,5.09,5.72
6,NaN,Arunachal Pradesh,105.13,105.09,105.12,2.48,2.80,2.56
7,NaN,Assam,106.03,104.45,105.71,3.74,2.53,3.49
8,NaN,Bihar,107.01,105.86,106.85,3.72,2.84,3.59


In [15]:
cpi.tail()

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
35,NaN,The Dadra And Nagar Haveli And Daman And Diu,108.00,108.05,108.04,3.49,4.22,4.03
36,NaN,Tripura,104.55,104.75,104.59,2.11,3.04,2.32
37,NaN,Uttar Pradesh,108.92,107.58,108.50,5.25,3.62,4.74
38,NaN,Uttarakhand,109.41,106.75,108.44,4.66,3.43,4.21
39,NaN,West Bengal,107.39,106.76,107.11,4.20,3.50,3.89


In [16]:
cpi.info()

<class 'pandas.DataFrame'>
RangeIndex: 36 entries, 4 to 39
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   s_no                0 non-null      float64
 1   state               36 non-null     str    
 2   cpi_rural           36 non-null     float64
 3   cpi_urban           36 non-null     float64
 4   cpi_combined        36 non-null     float64
 5   inflation_rural     36 non-null     float64
 6   inflation_urban     36 non-null     float64
 7   inflation_combined  36 non-null     float64
dtypes: float64(7), str(1)
memory usage: 2.4 KB


## Save the cleaned data as a CSV

In [17]:
cpi.to_csv("cpi_july_2026_cleaned.csv", index=False)

## Read the cleaned CSV back into Pandas

In [18]:
cpi_clean = pd.read_csv("cpi_july_2026_cleaned.csv")
cpi_clean.head()

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
0,NaN,Andaman And Nicobar Islands,111.13,107.44,109.32,4.99,3.26,4.15
1,NaN,Andhra Pradesh,107.81,107.55,107.71,6.13,5.09,5.72
2,NaN,Arunachal Pradesh,105.13,105.09,105.12,2.48,2.80,2.56
3,NaN,Assam,106.03,104.45,105.71,3.74,2.53,3.49
4,NaN,Bihar,107.01,105.86,106.85,3.72,2.84,3.59


In [19]:
cpi_clean.shape

(36, 8)

## Basic statistics

In [20]:
cpi_clean.describe()

,s_no,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
count,0.0,36.000000,36.000000,36.000000,36.000000,36.000000,36.000000
mean,NaN,104.736389,106.852500,107.370556,4.154444,3.620000,3.978056
std,NaN,18.037295,1.515066,1.529779,1.354359,0.956485,1.021519
min,NaN,0.000000,103.600000,104.230000,0.000000,1.520000,1.840000
25%,NaN,106.210000,105.472500,106.170000,3.442500,2.835000,3.365000
50%,NaN,107.740000,107.165000,107.785000,4.210000,3.580000,4.010000
75%,NaN,108.897500,107.942500,108.432500,5.120000,4.220000,4.760000
max,NaN,111.130000,109.690000,110.060000,6.860000,5.920000,6.320000


### Questions
- What is the average combined CPI?
- What is the average combined inflation rate?
- What is the minimum combined inflation rate?
- What is the maximum combined inflation rate?
- Which states appear to have relatively high inflation?

In [21]:
cpi_clean[["state", "cpi_combined", "inflation_combined"]]

,state,cpi_combined,inflation_combined
0,Andaman And Nicobar Islands,109.32,4.15
1,Andhra Pradesh,107.71,5.72
2,Arunachal Pradesh,105.12,2.56
3,Assam,105.71,3.49
4,Bihar,106.85,3.59
5,Chandigarh*,108.04,4.30
6,Chhattisgarh,105.59,3.37
7,Goa,106.29,3.35
8,Gujarat,107.23,4.01
9,Haryana,108.02,3.83


## Sort states by combined inflation

In [22]:
cpi_clean.sort_values(by="inflation_combined", ascending=False).head(10)

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
30,NaN,Telangana,110.56,109.69,110.06,6.86,5.92,6.32
1,NaN,Andhra Pradesh,107.81,107.55,107.71,6.13,5.09,5.72
29,NaN,Tamil Nadu,108.81,108.74,108.77,5.59,5.31,5.44
25,NaN,Puducherry,109.56,109.09,109.24,5.76,5.26,5.41
15,NaN,Ladakh,108.88,106.88,108.38,5.55,4.03,5.17
17,NaN,Madhya Pradesh,108.77,107.91,108.43,5.12,4.58,4.91
13,NaN,Karnataka,108.99,107.85,108.41,5.42,4.37,4.89
24,NaN,Odisha,107.35,106.28,107.06,5.12,4.02,4.83
27,NaN,Rajasthan,109.15,107.89,108.70,5.36,3.84,4.82
33,NaN,Uttar Pradesh,108.92,107.58,108.50,5.25,3.62,4.74


## Find the states with the lowest inflation

In [23]:
cpi_clean.sort_values(by="inflation_combined", ascending=True).head(10)

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined
21,NaN,Mizoram,105.07,103.60,104.23,2.27,1.52,1.84
20,NaN,Meghalaya,105.90,104.94,105.65,2.31,2.33,2.31
32,NaN,Tripura,104.55,104.75,104.59,2.11,3.04,2.32
2,NaN,Arunachal Pradesh,105.13,105.09,105.12,2.48,2.80,2.56
23,NaN,NCT of Delhi,105.89,105.36,105.38,4.12,2.65,2.68
22,NaN,Nagaland,106.61,104.52,105.81,2.82,2.82,2.82
19,NaN,Manipur,104.14,104.92,104.39,3.42,2.40,3.09
16,NaN,Lakshadweep,106.95,108.30,107.93,3.40,3.16,3.22
7,NaN,Goa,106.27,106.30,106.29,2.87,3.62,3.35
6,NaN,Chhattisgarh,105.88,105.04,105.59,3.84,2.46,3.37


## Filter states with inflation above 5%

In [24]:
high_inflation = cpi_clean[cpi_clean["inflation_combined"] > 5]
high_inflation[["state", "inflation_combined"]].sort_values(by="inflation_combined", ascending=False)

,state,inflation_combined
30,Telangana,6.32
1,Andhra Pradesh,5.72
29,Tamil Nadu,5.44
25,Puducherry,5.41
15,Ladakh,5.17


##  Filter states with inflation below 3%

In [25]:
low_inflation = cpi_clean[cpi_clean["inflation_combined"] < 3]
low_inflation[["state", "inflation_combined"]].sort_values(by="inflation_combined")

,state,inflation_combined
21,Mizoram,1.84
20,Meghalaya,2.31
32,Tripura,2.32
2,Arunachal Pradesh,2.56
23,NCT of Delhi,2.68
22,Nagaland,2.82


##  Calculate the rural-urban inflation gap

A positive value means rural inflation is higher than urban inflation. A negative value means urban inflation is higher.

In [26]:
cpi_clean["rural_urban_inflation_gap"] = cpi_clean["inflation_rural"] - cpi_clean["inflation_urban"]
cpi_clean[["state", "inflation_rural", "inflation_urban", "rural_urban_inflation_gap"]].head()

,state,inflation_rural,inflation_urban,rural_urban_inflation_gap
0,Andaman And Nicobar Islands,4.99,3.26,1.73
1,Andhra Pradesh,6.13,5.09,1.04
2,Arunachal Pradesh,2.48,2.80,-0.32
3,Assam,3.74,2.53,1.21
4,Bihar,3.72,2.84,0.88


## Find the largest rural-urban inflation gaps

In [27]:
cpi_clean[["state", "inflation_rural", "inflation_urban", "rural_urban_inflation_gap"]].sort_values(by="rural_urban_inflation_gap", ascending=False).head(10)

,state,inflation_rural,inflation_urban,rural_urban_inflation_gap
0,Andaman And Nicobar Islands,4.99,3.26,1.73
14,Kerala,4.91,3.27,1.64
33,Uttar Pradesh,5.25,3.62,1.63
27,Rajasthan,5.36,3.84,1.52
15,Ladakh,5.55,4.03,1.52
23,NCT of Delhi,4.12,2.65,1.47
18,Maharashtra,4.95,3.54,1.41
6,Chhattisgarh,3.84,2.46,1.38
34,Uttarakhand,4.66,3.43,1.23
3,Assam,3.74,2.53,1.21


## Find where urban inflation is higher

In [28]:
cpi_clean[["state", "inflation_rural", "inflation_urban", "rural_urban_inflation_gap"]].sort_values(by="rural_urban_inflation_gap", ascending=True).head(10)

,state,inflation_rural,inflation_urban,rural_urban_inflation_gap
5,Chandigarh*,0.00,4.30,-4.30
10,Himachal Pradesh,3.45,4.63,-1.18
32,Tripura,2.11,3.04,-0.93
7,Goa,2.87,3.62,-0.75
31,The Dadra And Nagar Haveli And Daman And Diu,3.49,4.22,-0.73
11,Jammu And Kashmir,3.74,4.22,-0.48
2,Arunachal Pradesh,2.48,2.80,-0.32
20,Meghalaya,2.31,2.33,-0.02
22,Nagaland,2.82,2.82,0.00
16,Lakshadweep,3.40,3.16,0.24


##  Compare rural and urban averages

In [29]:
print("Average rural inflation:", round(cpi_clean["inflation_rural"].mean(), 2))
print("Average urban inflation:", round(cpi_clean["inflation_urban"].mean(), 2))
print("Average combined inflation:", round(cpi_clean["inflation_combined"].mean(), 2))

Average rural inflation: 4.15
Average urban inflation: 3.62
Average combined inflation: 3.98


##  Compare Tamil Nadu with the rest of the dataset

In [30]:
tamil_nadu = cpi_clean[cpi_clean["state"].str.contains("Tamil Nadu", case=False, na=False)]
tamil_nadu

,s_no,state,cpi_rural,cpi_urban,cpi_combined,inflation_rural,inflation_urban,inflation_combined,rural_urban_inflation_gap
29,NaN,Tamil Nadu,108.81,108.74,108.77,5.59,5.31,5.44,0.28


In [31]:
print("Tamil Nadu combined CPI:", tamil_nadu["cpi_combined"].iloc[0])
print("Tamil Nadu rural inflation:", tamil_nadu["inflation_rural"].iloc[0])
print("Tamil Nadu urban inflation:", tamil_nadu["inflation_urban"].iloc[0])
print("Tamil Nadu combined inflation:", tamil_nadu["inflation_combined"].iloc[0])

Tamil Nadu combined CPI: 108.77
Tamil Nadu rural inflation: 5.59
Tamil Nadu urban inflation: 5.31
Tamil Nadu combined inflation: 5.44


##  Rank states by combined inflation

In [32]:
cpi_clean["inflation_rank"] = cpi_clean["inflation_combined"].rank(method="min", ascending=False)
cpi_clean[["state", "inflation_combined", "inflation_rank"]].sort_values("inflation_rank").head(10)

,state,inflation_combined,inflation_rank
30,Telangana,6.32,1.0
1,Andhra Pradesh,5.72,2.0
29,Tamil Nadu,5.44,3.0
25,Puducherry,5.41,4.0
15,Ladakh,5.17,5.0
17,Madhya Pradesh,4.91,6.0
13,Karnataka,4.89,7.0
24,Odisha,4.83,8.0
27,Rajasthan,4.82,9.0
33,Uttar Pradesh,4.74,10.0


##  Identify states above the average inflation rate

In [33]:
average_inflation = cpi_clean["inflation_combined"].mean()
above_average = cpi_clean[cpi_clean["inflation_combined"] > average_inflation]
print("Average combined inflation:", round(average_inflation, 2))
above_average[["state", "inflation_combined"]].sort_values("inflation_combined", ascending=False)

Average combined inflation: 3.98


,state,inflation_combined
30,Telangana,6.32
1,Andhra Pradesh,5.72
29,Tamil Nadu,5.44
25,Puducherry,5.41
15,Ladakh,5.17
17,Madhya Pradesh,4.91
13,Karnataka,4.89
24,Odisha,4.83
27,Rajasthan,4.82
33,Uttar Pradesh,4.74


##  Final data-quality check

In [34]:
print("Rows:", cpi_clean.shape[0])
print("Columns:", cpi_clean.shape[1])
print("Duplicate rows:", cpi_clean.duplicated().sum())
print("Missing values:")
print(cpi_clean.isnull().sum())

Rows: 36
Columns: 10
Duplicate rows: 0
Missing values:
s_no                         36
state                         0
cpi_rural                     0
cpi_urban                     0
cpi_combined                  0
inflation_rural               0
inflation_urban               0
inflation_combined            0
rural_urban_inflation_gap     0
inflation_rank                0
dtype: int64


## 26. Save the final cleaned dataset

In [35]:
cpi_clean.to_csv("cpi_july_2026_cleaned_final.csv", index=False)

# Story Pitch


## Tamil Nadu Rural Inflation Outpaces Urban Rate in July

Tamil Nadu's combined retail inflation stood at **5.44% in July 2026**, with rural inflation at **5.59%**, compared with **5.31% in urban areas**. The 0.28-percentage-point gap puts Tamil Nadu among the states with the highest inflation rates in the country and raises questions about what is driving price pressures outside the state's urban centres.

### Key Data Points

* **Tamil Nadu rural inflation:** 5.59%
* **Tamil Nadu urban inflation:** 5.31%
* **Rural-urban gap:** 0.28 percentage points
* **Combined inflation:** 5.44%
* **Rural CPI index:** 108.81
* **Urban CPI index:** 108.74
* **Combined CPI index:** 108.77
* **National combined inflation:** 3.98%
* **Tamil Nadu's position:** 3rd highest combined inflation among 36 states and Union Territories

### Story Angle

The July CPI data shows that **rural Tamil Nadu faced higher inflation than urban Tamil Nadu**, even though the rural and urban CPI indices were almost identical.

The data itself does not explain the gap. That creates the reporting question: **what is causing rural price pressures to run ahead of urban inflation?**

Possible explanations to investigate include differences in food prices, fuel costs, distribution and transportation costs, and changes in the prices of goods that make up rural household consumption baskets. These should be tested against category-level data and expert interviews rather than assumed from the headline numbers.

### A sustained rural-urban inflation gap could have implications for household purchasing power, particularly for agricultural workers and other rural households whose incomes may not rise as quickly as consumer prices. It could also provide a more granular picture of inflation in Tamil Nadu than the headline state-level number: the average inflation rate may mask different price experiences across rural and urban households.

### Stakeholders

* Tamil Nadu government/statistics officials for category-level CPI data
* Economists specialising in inflation and rural consumption
* Agricultural economists for food-price trends
* Retailers and distributors in rural Tamil Nadu
* Rural households and agricultural workers
* Logistics/transport industry representatives

### Next Steps

1. Obtain category-wise rural and urban CPI data for Tamil Nadu.
2. Compare July's rural-urban gap with the previous six months.
3. Compare Tamil Nadu with other states/ southern states.
4. Identify which consumption categories contributed most to the difference.
5. Speak to economists to see whether the gap is statistically or economically significant.
6. Speak to rural consumers and retailers to test whether the data reflects their experience of rising prices.
